### DISCUSSION TITLE
pdfium load: flatten  before load to support forms

### DISCUSSION THREAD

**Discussion by revathi-murali** [libvips/libvips#2034] [Ideas]
I have a sample pdf form where in I have added some info in the fillable areas. But when this pdf is converted to png, the info added in those fields are not present in the converted png.

**Input pdf:**
[lost_receipt_memo.pdf](https://github.com/libvips/libvips/files/3852764/lost_receipt_memo.pdf)

**Output png**

![response](https://user-images.githubusercontent.com/2732287/68966896-fc239a00-0793-11ea-87b9-b70655c6f8d4.png)

The above png does not have description of expenses or vendor name that is added in the pdf..

libvips version - 8.8.3 

We are using `pdfium` to convert from pdf to png

**Comment by jcupitt** [comment]
Sorry, pdfium does not support forms. Try viewing the PDF in Chrome -- you'll see the same problem.

**Comment by revathi-murali** [comment]
If you have a look at the discussion [here](https://bugs.chromium.org/p/pdfium/issues/detail?id=932), it says that by flattening the pdf, we will be able to render the forms properly.

So is there any roadmap for implementing that here?

**Comment by jcupitt** [comment]
Huh interesting. OK, let's tag this as an enhancement.

I don't have time to work on this right now, but perhaps someone will volunteer.

**Comment by zhaohuxing** [comment]
Hi @jcupitt, Now `libvips` supported that watermark for PDF?

**Comment by jcupitt** [comment]
You can watermark any image, but as a raster, of course. You can't use libvips to edit a PDF and insert a watermark layer.

**Comment by zhaohuxing** [comment]
Thanks @jcupitt 
BTW, Will `libvips` support PDF watermarks and other processing in the future?

**Comment by zhaohuxing** [comment]
@jcupitt  What PDF file processing does `libvips` support now? I didn't find relevant documents

**Comment by jcupitt** [comment]
libvips is an image processing library, it will not support PDF manipulation.

You can load PDFs as raster images and process them in any way you like. The PDF loader is documented here:

https://libvips.github.io/libvips/API/current/VipsForeignSave.html#vips-pdfload

**Comment by zhaohuxing** [comment]
Thanks
